frontend/src/components/data-table/cell-selection/feature.ts

```diff
@@ -0,0 +1,130 @@
+/* Copyright 2024 Marimo. All rights reserved. */
+"use no memo";
+
+import {
+  type Table,
+  type Cell,
+  type TableFeature,
+  type InitialTableState,
+  type RowData,
+  makeStateUpdater,
+  type Column,
+  type Row,
+} from "@tanstack/react-table";
+
+import { Functions } from "@/utils/functions";
+
+import type {
+  CellSelectionOptions,
+  CellSelectionState,
+  CellSelectionTableState,
+} from "./types";
+import { INDEX_COLUMN_NAME } from "../types";
+
+function getRowId<TData>(row: Row<TData>): string {
+  if (row && typeof row === "object" && INDEX_COLUMN_NAME in row) {
+    return String(row[INDEX_COLUMN_NAME]);
+  }
+  return row.id;
+}
+
+export const CellSelectionFeature: TableFeature = {
+  getInitialState: (state?: InitialTableState): CellSelectionTableState => {
+    return {
+      ...state,
+      cellSelection: [] as CellSelectionState,
+    };
+  },
+
+  getDefaultOptions: <TData extends RowData>(
+    table: Table<TData>,
+  ): CellSelectionOptions => ({
+    onCellSelectionChange: makeStateUpdater("cellSelection", table),
+    enableCellSelection: false,
+  }),
+
+  createTable: <TData>(table: Table<TData>): void => {
+    table.setCellSelection = (updater) => {
+      // TODO: can we access?
+      // table._getRowId
+      // check if pagination is active
+
+      table.setState((tableState) => ({
+        ...tableState,
+        cellSelection: Functions.asUpdater(updater)(tableState.cellSelection),
+      }));
+      table.options.onCellSelectionChange?.(updater);
+    };
+
+    table.resetCellSelection = (defaultValue) => {
+      table.setCellSelection(() => defaultValue ?? []);
+    };
+
+    table.getIsAllCellsSelected = () => {
+      const state = table.getState().cellSelection ?? [];
+      return (
+        state.length === table.getRowCount() * table.getAllColumns().length
+      );
+    };
+
+    table.getIsAllPageCellsSelected = () => false; // TODO: I don't quite have a notion of pages yet
```

Comment:
i think we can remove this. i think its when you use a checkbox at the top of the table to select all within the page, so i don't think itll make sense